### PR TITLE
fix(vm): set power state of VM centrally

### LINKED ISSUE #2284: vm: Modifying a stopped VM which's state should be started hangs on apply
**Describe the bug**
I noticed while testing #2260 that when a VM is stopped but should be started and I modify it (e.g. remove network devies as in this case), the apply hangs because the provider sends a "start" command to the VM (it was stopped) and immediately a "reboot" signal afterwards. This is a) not required, since the change was already applied when the VM was stopped and b) doesn't work as the VM doesn't receive the reboot signal and the apply just times out.

I think it should be detected that the VM is turned off (already known) and the reboot is not sent in this case, but only a start command.

**To Reproduce**

1. Create a config like the one in the linked PR
2. apply state
3. Turn off VM in proxmox
4. Delete network interface in opentofu config
5. apply state
6. See timeout and pending proxmox reboot task

**Expected behavior**
The reboot is not sent and the apply succeeds after starting the VM.

- Single or clustered Proxmox:
- Proxmox version: 9.0.10
- Provider version (ideally it should be the latest version): 0.85.1
- Terraform/OpenTofu version: 1.10.6
- OS (where you run Terraform/OpenTofu from): OpenSUSE Tumbleweed

### PR DESCRIPTION
**Pull request by TobiPeterG**
### What does this PR do?

Instead of each function starting/stopping VMs, we have centralized functions now that track our changes. In the end, one final function makes sure that the VM is in the desired state after all actions ran. This also refactors the disk.Update() function to not return an always false parameter anymore.

### ⚠ BREAKING CHANGES

- `proxmox_virtual_environment_vm`: `template` attribute no longer forces resource recreation — conversion now happens in-place                                                                                                                                
- `proxmox_virtual_environment_vm`: operations requiring VM shutdown (TPM, cloud-init move, template conversion, disk move) now fail when `reboot_after_update = false` instead of silently powering off the VM

### Contributor's Note
- [X] I have run `make lint` and fixed any issues.
- [x] I have updated documentation (FWK: schema descriptions + `make docs`; SDK: manual `/docs/` edits).
- [X] I have added / updated acceptance tests for any new or updated resources / data sources.
- [x] I have run `make example` to verify the change works (mainly for SDK / provider config changes).

### Proof of Work
I don't have a seperate proxmox instance to test this currently, so I can't promise that it will work. However, I already added a test case where I'm fairly certain that the current version will fail. I will update this as soon as I ran the tests and example folder. :)

### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #2284

### CODE REVIEW

File: proxmoxtf/resource/vm/vm.go [hunk 1]
```diff
@@ -6488,54 +6616,6 @@ func vmUpdateDiskLocationAndSize(
...
-				}
-
-				if diags := vmStart(ctx, vmAPI, d); diags != nil {
-					return diags
-				}
-			}
-		}
 	}
 
 	return vmRead(ctx, d, m)
```
[the remark this answers is not in the document]
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: Done

File: proxmoxtf/resource/vm/vm.go [hunk 2]
```diff
@@ -1967,6 +1967,145 @@ func vmStop(ctx context.Context, vmAPI *vms.Client, d *schema.ResourceData) diag
...
+		} else {
+			if diags := vmStop(ctx, vmAPI, d); diags != nil {
+				return diags
+			}
+
+			power.stoppedByProvider = true
+
+			if diags := vmStart(ctx, vmAPI, d); diags != nil {
+				return diags
+			}
```
[the remark this answers is not in the document]
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: I think this is not valid right? We don't want to just turn the VM off or on, we want to reboot it. We check if the agent is enabled and if yes, perform a regular reboot. However, if not, we know that we can't use e.g. the shutdown or reboot command, so we have to use the stop command and a start afterwards. Of course, this could use EnsureStopped and EnsureRunning, but EnsureStopped would come to the same conclusion that only vmStop is possible and EnsureStarted would see that the VM is stopped and call vmStart. The result would be the same, but with additional checks in between. I guess the performance loss/time loss is very small, but this is still the slightly optimized version. 
Other opinions?

File: proxmoxtf/resource/vm/vm.go [hunk 3]
```diff
@@ -1967,6 +1967,145 @@ func vmStop(ctx context.Context, vmAPI *vms.Client, d *schema.ResourceData) diag
...
+	if startedDesired {
+		if currentlyStopped {
+			if diags := vmStart(ctx, vmAPI, d); diags != nil {
+				return diags
+			}
+
+			return nil
+		}
+	} else {
+		if (startedChanged) && !currentlyStopped {
```
[the remark this answers is not in the document]
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: Done

File: proxmoxtf/resource/vm/vm.go
Comment (bpg-dev) [inline proxmoxtf/resource/vm/vm.go]: not your changes, but spotted this
```suggestion
			return diag.FromErr(er)
```

File: fwprovider/test/resource_vm_test.go [hunk 4]
```diff
@@ -1663,3 +1663,71 @@ func TestAccResourceVMVirtioSCSISingleWithAgent(t *testing.T) {
...
+			{
+				PreConfig: func() {
+					if vmID == "" {
+						t.Fatalf("vm_id was not captured from state")
+					}
+					err := te.StopVM(te.NodeName, vmID)
+					if err != nil {
+						t.Fatalf("failed to stop VM out-of-band: %v", err)
+					}
+					te.WaitForVMStopped(te.NodeName, vmID)
```
Comment (bpg-dev) [inline fwprovider/test/resource_vm_test.go]: i guess, test environment's changes are missing from this PR, `te.StopVM|te.WaitForVMStopped` are not there.

File: proxmoxtf/resource/vm/vm.go [hunk 5]
```diff
@@ -6265,19 +6381,32 @@ func vmUpdate(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnosti
...
-		m,
+		template,
+		started,
 		!template && rebootRequired,
-	)
+		power,
+		d.HasChange(mkStarted),
+	); diags != nil {
+		return diags
+	}
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm/vm.go:6708]: here and in other places, `diags` is not an equivalent of `err`, it may contain warnings as well. So if 'finalize' returned a waring diagnostic (when `eboot_after_update = false`) it will exit without reading the state.

prob. return errors from the helpers and convert them to diags as needed? Or check for `diags.HasError()` instead of nil

File: fwprovider/test/resource_vm_test.go [hunk 6]
```diff
@@ -1663,3 +1663,71 @@ func TestAccResourceVMVirtioSCSISingleWithAgent(t *testing.T) {
...
+							cores = 2
+							type  = "host"
+						}
+
+						memory {
+							dedicated = 1024
+						}
+					}
+				`),
+			},
```
Comment (bpg-dev) [inline fwprovider/test/resource_vm_test.go:1961]: should prob. add a check if VM is running

File: proxmoxtf/resource/vm/vm.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bpg-dev) [reply proxmoxtf/resource/vm/vm.go]: it seems ok, but kinda micro-optimizations. if we make all start/stop checks look consistent it will be better IMO

**Review by bpg — COMMENTED** [review]
not sure if this code is ready for review, but added few comments

**Comment by TobiPeterG** [comment]
Currently blocked because of #2578

**Comment by TobiPeterG** [comment]
> not sure if this code is ready for review, but added few comments

Thanks, I can't tell if it's ready because I can't execute the exmaples or acceptance tests due to #2578 

But thanks for the initial comments, once I have a working test setup I can fix occuring issues and work on your comments :)

**Comment by bpg-dev** [comment]
Thank you @TobiPeterG. I've commented on the #2578 ticket, hope it makes sense you can unblock your test setup.
Let me know if you have any more troubles with acceptance tests or running examples, i can re-test your branch in my lab and give you feedback. Will be testing it before merge anyway :)

**Comment by TobiPeterG** [comment]
> Thank you @TobiPeterG. I've commented on the #2578 ticket, hope it makes sense you can unblock your test setup. Let me know if you have any more troubles with acceptance tests or running examples, i can re-test your branch in my lab and give you feedback. Will be testing it before merge anyway :)

Thanks a lot, I can run my acceptance test now. It currently fails because my VM doesn't have an IP address and therefore the enabled agent isn't happy (but this is required for the test) and with the main version, I can see that there is the stale reboot command and with this branch there isn't :)

However, this was before rebasing. I will do so now :)

**Comment by TobiPeterG** [comment]
Ok, I rebased it. I moved the logic to restart the VM for pending disk changes a bit, I hope this is ok. I will fix my acceptance test now (add a cloud config for dhcp IP). This might take a bit for me to do since I need to add a DHCP server to my test network first :D  

However, @bpg-dev feel free to already take a look at the changes in general :)

EDIT: I will also incorportate your review comments with the next push :)

File: fwprovider/test/resource_vm_test.go (hunk 4 above)
Comment (TobiPeterG) [reply fwprovider/test/resource_vm_test.go]: Added them as general helpers :)

File: fwprovider/test/resource_vm_test.go (hunk 6 above)
Comment (TobiPeterG) [reply fwprovider/test/resource_vm_test.go:1961]: Should be done with next push, please verify then :)

File: proxmoxtf/resource/vm/vm.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: Should be fixed after next push :)

File: proxmoxtf/resource/vm/vm.go
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: Should be fixed with the next push :)

File: proxmoxtf/resource/vm/vm.go (hunk 5 above)
Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go:6708]: Should be fixed, please verify :)
(I hope I didn't miss a place :D )

File: proxmoxtf/resource/vm/vm.go [hunk 7]
```diff
@@ -6330,42 +6556,74 @@ func vmUpdate(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnosti
...
+
+	if pendingChangesPowerOff.handledOffline {
+		rebootRequired = false
+	}
+
+	if pendingChangesPowerOff.blockedByPolicy {
+		// Return here to avoid applying further modifications that could override other changes.
+		updateDiags = append(updateDiags, vmRead(ctx, d, m)...)
+		return updateDiags
+	}
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm/vm.go]: When `reboot_after_update = false` and a reboot is needed, the new code's early return here at `blockedByPolicy`, which prevents `vmUpdateDiskLocationAndSize` from being called at all, skipping both disk moves **and** disk resizes. The old code entered `vmUpdateDiskLocationAndSize` where disk moves (with their own shutdown/start cycle) were executed before the reboot check; only disk resizes were skipped. 

This looks like regression / behaviour change to me, and I want to make sure i understand the code correctly and this change is intentional.
If "yes" and "yes", then let's discuss why it is needed.
If it's not intentional, then I believe the code can be restructured so disk moves are still performed even when `blockedByPolicy` is true.

Comment (bpg-dev) [reply proxmoxtf/resource/vm/vm.go]: One more thing from the previous comments, it probably got lost in the weeds
The checks like `if diags != nil { return ... }` should really be `diags.HasError()` instead. The reason is `diags` is not a replacement of golang errors, they may include warnings as well, in which case we should continue the flow and not abort.

Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: > When `reboot_after_update = false` and a reboot is needed, the new code's early return here at `blockedByPolicy`, which prevents `vmUpdateDiskLocationAndSize` from being called at all, skipping both disk moves **and** disk resizes. The old code entered `vmUpdateDiskLocationAndSize` where disk moves (with their own shutdown/start cycle) were executed before the reboot check; only disk resizes were skipped.
> 
> This looks like regression / behaviour change to me, and I want to make sure i understand the code correctly and this change is intentional. If "yes" and "yes", then let's discuss why it is needed. If it's not intentional, then I believe the code can be restructured so disk moves are still performed even when `blockedByPolicy` is true.

I split the function to do size moves and resizes into two and apply the disk move operation before evaluating if a reboot is required. This should behave the same now

Comment (TobiPeterG) [reply proxmoxtf/resource/vm/vm.go]: > One more thing from the previous comments, it probably got lost in the weeds The checks like `if diags != nil { return ... }` should really be `diags.HasError()` instead. The reason is `diags` is not a replacement of golang errors, they may include warnings as well, in which case we should continue the flow and not abort.

Sorry, missed those, should hopefully be fixed. :)

**Comment by TobiPeterG** [comment]
@bpg-dev I reworked and rebased my code; I noticed that in vmUpdate, some parts just directly shutdown the VM to apply a change (e.g. TPM devices), which for me felt like violating the user setting that the VM shouldn't be rebooted. Therefore, I block those operations now when reboots are disallowed. This is a behavorial change. Should I document it somewhere? Is this even desired?

(make example also still runs through)

File: fwprovider/test/resource_vm_test.go [hunk 8]
```diff
@@ -1731,3 +1736,212 @@ func TestAccResourceVMVirtioSCSISingleWithAgent(t *testing.T) {
...
+	id, err := strconv.Atoi(vmID)
+	if err != nil {
+		return fmt.Errorf("invalid vm_id %q: %w", vmID, err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+
+	return te.NodeClient().VM(id).StopVM(ctx)
+}
```
[the remark this answers is not in the document]
Comment (TobiPeterG) [reply fwprovider/test/resource_vm_test.go:1976]: We create the VM powered off, so this won't fail.

**Comment by bpg-dev** [comment]
> @bpg-dev I reworked and rebased my code; I noticed that in vmUpdate, some parts just directly shutdown the VM to apply a change (e.g. TPM devices), which for me felt like violating the user setting that the VM shouldn't be rebooted. Therefore, I block those operations now when reboots are disallowed. This is a behavorial change. Should I document it somewhere? Is this even desired?
> 
> (make example also still runs through)

Hey @TobiPeterG 👋🏼 

This PR is too big so I can't fit it all in my head in time i have for code reviews, so I used LLM to help me.

But this change triggered my mental alarm regardless. The VM power off is required for certain things, like disk moves, and prob something else. It can't be optional in such cases, and I don't think [`reboot_after_update`](https://bpg.sh/docs/resources/virtual_environment_vm/#reboot-after-update) flag should be considered there, as if the VM was not reboot / powered off at exact moment, the change will be lost, and a manual reboot after the fact won't fix that.

In the hindsight, this flag is bad. It allows users to intervene in the flow and potentially break it without noticing, or unintentionally.

Below is the LLM's comment:

---

### `vmPowerOffForPendingChanges` is used for mandatory shutdowns that should not be gated by `reboot_after_update`

The `vmPowerOffForPendingChanges` function (line 2165) was designed for **optional reboot-after-config-change** scenarios — where a VM _could_ be rebooted to apply pending changes, but the user can opt out via `reboot_after_update = false`. That's the correct use for the disk-resize reboot path (line 6632).

However, this same function is now also used for four operations that require **unconditional** VM shutdown — these are not optional reboots, they are _prerequisites_ for the operation to succeed:

| Operation | Call site | Why shutdown is mandatory |
|-----------|-----------|--------------------------|
| TPM state change | Line 6171 | Proxmox requires VM stopped to modify TPM device |
| Cloud-init move | Line 6258 | Requires VM stopped to move cloud-init between interfaces/datastores |
| Template conversion | Line 6570 | `ConvertToTemplate` requires VM to be stopped first |
| Disk move | Line 6614 | Proxmox API cannot move disks while VM is running |

When `reboot_after_update = false`, `vmPowerOffForPendingChanges` returns `blockedByPolicy: true`, and all four call sites do an **early return** without performing the operation:

```go
if powerOffForTemplate.blockedByPolicy {
    diags := append(diag.Diagnostics{}, powerOffForTemplate.diags...)
    diags = append(diags, vmRead(ctx, d, m)...)
    return diags  // ← template conversion never happens
}
```

**Template conversion is the worst case** — `ConvertToTemplate` (line 6583) is never called, but `vmRead` refreshes state, so Terraform may show `template = true` while the actual VM is unchanged.

**Before this PR**, all four operations used unconditional shutdown (`vmShutdown` or `vmStop`) without consulting `reboot_after_update`. For example, the old template conversion code:

```go
// Old code (commit 4ffb2f18):
if status != nil && status.Status != "stopped" {
    if diags := vmStop(ctx, vmAPI, d); diags != nil {
        return diags
    }
}
e = vmAPI.ConvertToTemplate(ctx)  // Always reached
```

And the old TPM code:

```go
// Old code:
if !stoppedBeforeUpdate {
    if er := vmShutdown(ctx, vmAPI, d); er != nil {
        return er
    }
    stoppedBeforeUpdate = true
}
```

**Suggested fix:** Separate the two use cases. For mandatory shutdowns (TPM, cloud-init move, template conversion, disk move), call `power.EnsureStopped()` directly — these operations must stop the VM regardless of `reboot_after_update`. Reserve `vmPowerOffForPendingChanges` for the _optional_ reboot path (pending config changes before disk resize), which is the only case where `reboot_after_update` should be consulted.

This was also flagged in the previous review round — see @bpg-dev's comment from March 6 regarding `blockedByPolicy` skipping disk moves.

**Comment by TobiPeterG** [comment]
> > @bpg-dev I reworked and rebased my code; I noticed that in vmUpdate, some parts just directly shutdown the VM to apply a change (e.g. TPM devices), which for me felt like violating the user setting that the VM shouldn't be rebooted. Therefore, I block those operations now when reboots are disallowed. This is a behavorial change. Should I document it somewhere? Is this even desired?
> > (make example also still runs through)
> 
> Hey @TobiPeterG 👋🏼
> 
> This PR is too big so I can't fit it all in my head in time i have for code reviews, so I used LLM to help me.
> 
> But this change triggered my mental alarm regardless. The VM power off is required for certain things, like disk moves, and prob something else. It can't be optional in such cases, and I don't think [`reboot_after_update`](https://bpg.sh/docs/resources/virtual_environment_vm/#reboot-after-update) flag should be considered there, as if the VM was not reboot / powered off at exact moment, the change will be lost, and a manual reboot after the fact won't fix that.
> 
> In the hindsight, this flag is bad. It allows users to intervene in the flow and potentially break it without noticing, or unintentionally.
> 
> Below is the LLM's comment:
> 
> ### `vmPowerOffForPendingChanges` is used for mandatory shutdowns that should not be gated by `reboot_after_update`
> The `vmPowerOffForPendingChanges` function (line 2165) was designed for **optional reboot-after-config-change** scenarios — where a VM _could_ be rebooted to apply pending changes, but the user can opt out via `reboot_after_update = false`. That's the correct use for the disk-resize reboot path (line 6632).
> 
> However, this same function is now also used for four operations that require **unconditional** VM shutdown — these are not optional reboots, they are _prerequisites_ for the operation to succeed:
> 
> Operation	Call site	Why shutdown is mandatory
> TPM state change	Line 6171	Proxmox requires VM stopped to modify TPM device
> Cloud-init move	Line 6258	Requires VM stopped to move cloud-init between interfaces/datastores
> Template conversion	Line 6570	`ConvertToTemplate` requires VM to be stopped first
> Disk move	Line 6614	Proxmox API cannot move disks while VM is running
> When `reboot_after_update = false`, `vmPowerOffForPendingChanges` returns `blockedByPolicy: true`, and all four call sites do an **early return** without performing the operation:
> 
> ```go
> if powerOffForTemplate.blockedByPolicy {
>     diags := append(diag.Diagnostics{}, powerOffForTemplate.diags...)
>     diags = append(diags, vmRead(ctx, d, m)...)
>     return diags  // ← template conversion never happens
> }
> ```
> 
> **Template conversion is the worst case** — `ConvertToTemplate` (line 6583) is never called, but `vmRead` refreshes state, so Terraform may show `template = true` while the actual VM is unchanged.
> 
> **Before this PR**, all four operations used unconditional shutdown (`vmShutdown` or `vmStop`) without consulting `reboot_after_update`. For example, the old template conversion code:
> 
> ```go
> // Old code (commit 4ffb2f18):
> if status != nil && status.Status != "stopped" {
>     if diags := vmStop(ctx, vmAPI, d); diags != nil {
>         return diags
>     }
> }
> e = vmAPI.ConvertToTemplate(ctx)  // Always reached
> ```
> 
> And the old TPM code:
> 
> ```go
> // Old code:
> if !stoppedBeforeUpdate {
>     if er := vmShutdown(ctx, vmAPI, d); er != nil {
>         return er
>     }
>     stoppedBeforeUpdate = true
> }
> ```
> 
> **Suggested fix:** Separate the two use cases. For mandatory shutdowns (TPM, cloud-init move, template conversion, disk move), call `power.EnsureStopped()` directly — these operations must stop the VM regardless of `reboot_after_update`. Reserve `vmPowerOffForPendingChanges` for the _optional_ reboot path (pending config changes before disk resize), which is the only case where `reboot_after_update` should be consulted.
> 
> This was also flagged in the previous review round — see @bpg-dev's comment from March 6 regarding `blockedByPolicy` skipping disk moves.

Personally, I feel like the current implementation gives users a wrong sense of security when they set reboot_after_update to false. I would _think_ as a user that no matter what change I do, the VM will never reboot /  power off to perform a change automatically (and pwering the VM off, perform a change and turn it back on in the end is conceptually very close to a reboot, isn't it). What would you expect?

It is wanted that the four operations now return early - the change they are intended to do are not possible without intervening with the user not wanting the VM to automatically be taken offline. If the state is wrongly calculated afterwards, I'll look into this, but I think ignoring the users which to keep the VM online and just powering it off anyway feels wrong to me. I hope my thoughts make sense :)

**Comment by bpg-dev** [comment]
> Personally, I feel like the current implementation gives users a wrong sense of security when they set reboot_after_update to false. I would think as a user that no matter what change I do, the VM will never reboot / power off to perform a change automatically (and pwering the VM off, perform a change and turn it back on in the end is conceptually very close to a reboot, isn't it). What would you expect?

> It is wanted that the four operations now return early - the change they are intended to do are not possible without intervening with the user not wanting the VM to automatically be taken offline. If the state is wrongly calculated afterwards, I'll look into this, but I think ignoring the users which to keep the VM online and just powering it off anyway feels wrong to me. I hope my thoughts make sense :)

Yeah, i see you point. I guess you're coming from a perspective of managing production systems, my mindset is more of a homelabber :) 

So it looks like we have 2 tiers of things that require reboot during the VM update, and having a blanket "aplly, but not reboot" would be wrong for them. Things that can be fixed by a subsequent manual reboot, like a disk resize or updating a non-hotpluggable component. Then things that can't, like a disk move, and perhaps something else (AI suggests TPM, cloud-init move, template conversion -- probably it's right)
So for these cases we need to fail at `apply` (i don't think we can reasonable detect that and fail at `plan`), as there will be no recovery as simple as a reboot.

If your current code is already doing that -- awesome! I'd just need to have a proof of work in this PR for future selves dealing with this again when migrating VM resource to FWK 😅. If not, I'd like you to spend a bit more time on untangling that, and do some extra tests, manual or via acceptance tests.

I hope i made myself clear. But if you see any gaps in my logic or understanding of this changeset, please feel free to comment, i'll be happy to continue the discussion.

**Comment by TobiPeterG** [comment]
@bpg-dev 

thank you for your message :)

I reworked the code again to hard fail when an action cannot be applied without powering the VM off. I also added acceptance tests for the cases I identified where this is necessary.

One question: converting to a template was marked as ForceNew, which led to it not going through the vmUpdate path. Was this intentional? I removed it and all tests I executed ran through fine (though I didn't run all acceptance tests, "just" ./testacc TestAccResourceVM*, but this should hopefully cover all VM cases (see below for failing cases)). I also couldn't find any reasoning for why it's marked as ForceNew and a git blame also didn't help. Is it required? I removed it for now, but the change might now count as a breaking change (Should I mark it as breaking change?).

Please check if the new acceptance tests cover all relevant cases. :)


As for the acceptance tests: The only test failure I saw when running ./testacc TestAccResourceVM* was:
```
--- FAIL: TestAccResourceVMDisks (3.18s)
    --- FAIL: TestAccResourceVMDisks/clone_with_moving_disk (60.69s)
        resource_vm_disks_test.go:1416: Step 1/1 error: Error running apply: exit status 1
            
            Error: disk move fails: error moving VM disk: the requested resource does not exist
            received an HTTP 500 response - Reason: storage 'tank' does not exist
            
              with proxmox_virtual_environment_vm.clone,
              on terraform_plugin_test.tf line 38, in resource "proxmox_virtual_environment_vm" "clone":
              38:                               resource "proxmox_virtual_environment_vm" "clone" {
            
FAIL
FAIL    github.com/bpg/terraform-provider-proxmox/fwprovider/test       984.182s
```
And I'm wondering where the "tank" storage comes from. Did I miss documentation on how to set that up? My test environment only has local and local-lvm storage pools. Or is this a real regression?

**Comment by bpg-dev** [comment]
> @bpg-dev
> 
> thank you for your message :)
> 
> I reworked the code again to hard fail when an action cannot be applied without powering the VM off. I also added acceptance tests for the cases I identified where this is necessary.

Thank you! ❤️ 

> 
> One question: converting to a template was marked as ForceNew, which led to it not going through the vmUpdate path. Was this intentional? I removed it and all tests I executed ran through fine (though I didn't run all acceptance tests, "just" ./testacc TestAccResourceVM*, but this should hopefully cover all VM cases (see below for failing cases)). I also couldn't find any reasoning for why it's marked as ForceNew and a git blame also didn't help. Is it required? I removed it for now, but the change might now count as a breaking change (Should I mark it as breaking change?).
> 

That's hard to say why it's `ForceNew`. I guess for historical reasons, inherited from the legacy code.
I think removal of it makes sense, but it will be a breaking change in the behaviour, i'll add a note to the Release Notes.

> Please check if the new acceptance tests cover all relevant cases. :)

Looks good so far, but i'll do more testing before the release.

> 
> As for the acceptance tests: The only test failure I saw when running ./testacc TestAccResourceVM* was:
> 
> ```
> --- FAIL: TestAccResourceVMDisks (3.18s)
>     --- FAIL: TestAccResourceVMDisks/clone_with_moving_disk (60.69s)
>         resource_vm_disks_test.go:1416: Step 1/1 error: Error running apply: exit status 1
>             
>             Error: disk move fails: error moving VM disk: the requested resource does not exist
>             received an HTTP 500 response - Reason: storage 'tank' does not exist
>             
>               with proxmox_virtual_environment_vm.clone,
>               on terraform_plugin_test.tf line 38, in resource "proxmox_virtual_environment_vm" "clone":
>               38:                               resource "proxmox_virtual_environment_vm" "clone" {
>             
> FAIL
> FAIL    github.com/bpg/terraform-provider-proxmox/fwprovider/test       984.182s
> ```
> 
> And I'm wondering where the "tank" storage comes from. Did I miss documentation on how to set that up? My test environment only has local and local-lvm storage pools. Or is this a real regression?

The `tank` storage is an oversight, should document it. Added this test for validating edge cases with ZFS pools. This test works for me on your branch.
